src/XrdTpc/PMarkManager.hh

```diff
@@ -0,0 +1,113 @@
+#include "XrdNet/XrdNetAddrInfo.hh"
+#include "XrdNet/XrdNetAddr.hh"
+#include "XrdHttp/XrdHttpExtHandler.hh"
+
+#include <map>
+#include <memory>
+#include <queue>
+
+/**
+ * This class will manage packet marking handles for TPC transfers
+ *
+ * Each time a socket will be opened by curl (via the opensocket_callback), the manager
+ * will register the related information to the socket.
+ *
+ * Once the transfer will start we will start the packet marking by creating XrdNetPMark::Handle
+ * objects from the socket information previously registered.
+ *
+ * In the case of multi-stream HTTP TPC transfers, a packet marking handle will be created for each stream.
+ * The first one will be created as a basic one. The other will be created using the first packet marking handle as a basis.
+ */
+class PMarkManager {
+public:
+
+  /**
+   * This class allows to create and keep a XrdSecEntity object
+   * from the socket file descriptor and address
+   * Everything is done on the constructor
+   *
+   * These infos will be used later on when we create new PMark handles
+   */
+  class SocketInfo {
+  public:
+    SocketInfo(int fd, const struct sockaddr * sockP);
+    XrdNetAddr netAddr;
+    XrdSecEntity client;
+  };
+
+  PMarkManager(XrdNetPMark * pmark);
+  /**
+   * Add the connected socket information that will be used for packet marking to this manager class
+   * Note: these info will only be added if startTransfer(...) has been called. It allows
+   * to ensure that the connection will be related to the data transfers and not for anything else. We only want
+   * to mark the traffic of the transfers.
+   * @param fd the socket file descriptor
+   * @param sockP the structure describing the address of the socket
+   */
+  void addFd(int fd, const struct sockaddr * sockP);
+
+  /**
+   * Calling this function will indicate that the connections that will happen will be related to the
+   * data transfer. The addFd(...) function will then register any socket that is created after this function
+   * will be called.
+   * @param req the request object that will be used later on to get some information about the transfer
+   */
+  void startTransfer(XrdHttpExtReq * req);
+  /**
+   * Creates the different packet marking handles allowing to mark the transfer packets
+   *
+   * Call this after the curl_multi_perform() has been called.
+   */
+  void beginPMarks();
+
+  /**
+   * This function deletes the PMark handle associated to the fd passed in parameter
+   * Use this before closing the associated socket! Otherwise the information contained in the firefly
+   * (e.g sent bytes or received bytes) will have values equal to 0.
+   * @param fd the fd of the socket to be closed
+   */
+  void endPmark(int fd);
+
+  virtual ~PMarkManager() = default;
+private:
+  // The list of socket information from which we will create the packet marking handles
+  std::queue<SocketInfo> mSocketInfos;
+  // The vector of the packet marking handles
```

Comment:
You may want to update also this comment.

Reply:
Yes haha, done :)